src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java

```diff
@@ -3828,6 +3828,51 @@ private List<Type> pruneInterfaces(Type t) {
         return result.toList();
     }
 
+    /**
+     * Determine if an instance field may appear in a constructor prologue.
+     *
+     * <p>
+     * This is only allowed when:
+     *  - The field is being assigned a value (i.e., written but not read)
+     *  - The field is not inherited from a superclass
+     *  - The assignment is not within a lambda, because that would require
+     *    capturing 'this' which is not allowed prior to super().
+     *
+     * <p>
+     * Note, this method doesn't catch all such scenarios, because this method
+     * is invoked for symbol "x" only for "x = 42" but not for "this.x = 42".
+     * To catch those cases, we rely on similar logic in Attr.checkAssignable().
+     */
+    private boolean isAllowedInPrologue(Env<AttrContext> env, VarSymbol v) {
+
+        // Check assumptions
+        Assert.check(env.info.ctorPrologue);
+        Assert.check((v.flags() & STATIC) == 0);
```

Comment:
nit: I would prefer here to directly read the `flags_field` field instead of invoking `Symbol::flags` as the later can provoke symbol completions. Probably not a big deal here but just to be on the safe side

Reply:
Thanks - I didn't realize there could be side effects, and it's especially appropriate to avoid them because this is just an assertion statement.

Fixed in 34aa494680f.